azero/contracts/ownable/lib.rs

```diff
@@ -0,0 +1,129 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+/// This crate facilitates introducing the ownership of contracts and ownership changes using a two-step process.
+///
+/// The basic functionality is similar to the `Ownable` concept and exposes the following messages/methods:
+/// * `get_owner`
+/// * `is_owner`
+/// * `ensure_owner`: a helper to use with the `?` syntax that will check whether the caller is the owner of the contract
+///
+/// Additionally, it introduces the following method for transferring ownership:
+/// * `transfer_ownership`: callable only by the current owner, appoints the new owner but instead of making them the owner right away, it stores them in the `pending_owner` field
+/// * `accept_owership`: callable only by the pending owner, removes the previous owner and makes them the sole owner of the contract
+/// * `get_pending_owner`: returns the pending owner, if the ownership change process is currently underway.  
+///
+/// In order to use it in your contract, implement the methods of the `Ownable2Step` trait: in most cases, you can simply call the corresponding methods on the `Data` object.
+use ink::primitives::AccountId;
+use scale::{Decode, Encode};
+
+#[derive(Debug, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum Error {
+    /// The caller didn't have the permissions to call a given method
+    UnauthorizedAccount(AccountId),
+    /// The caller tried to accept ownership but the process hasn't been started
+    NoPendingOwner,
+    /// Useful in cases, when the `Data` struct is not accessed directly but inside of `Lazy` or a `Mapping`, means that we failed to access the `Data` struct itself.
+    CorruptedStorage,
```

Comment:
I don't think it should be a part of this Error enum, as it's not an error originating from this struct, but from the contract that uses it

Reply:
I agree. It's been done this way because `From<MostError>` for `Ownable2Step::Error` cannot be implemented in the Most contract since it's in the external crate. I'll think about more sleek solution.

Reply:
Anyway, implementing `From<MostError>` for `ownable2Step::Error` would cause recursive types (as `Form<ownable2Step::Error>` has been implemented for `MostError` too) 🤔.
- Should we abandon `OwnableResult<T>` including `ownable2Step::Error`, and return booleans instead of `Return<(),ownable2step::Error>` were appropriate?
Or
- Replace `ownable2step::Error::CorruptedStorage` with more generic `ownabl2step::Error::Custom(String)`... ?

@piotrMocz

Reply:
I'd be a fan of the second option, to stay within the `Result` state of mind ;) It's kinda annyoying and a bit backwards that the implementing contract influences the way we design the trait but first of all, we're writing this library specifically for this contract and secondly, this is mostly due to limitations of ink traits, so it's what we've got to work with

Reply:
Unfortunately, you won't escape it (i.e. that contract influences the `trait`). So I stopped treating contract's traits as traits in "normal programming" and treat them as "whole package" and extract the trait only to minimise the dependencies later on.